compose/snippets/src/main/java/com/example/compose/snippets/pictureInPicture/PictureInPictureSnippets.kt

```diff
@@ -0,0 +1,384 @@
+            onDispose {
+            }
+        }
+    }
+}
+
+@Composable
+fun VideoPlayer1(
+    shouldEnterPipMode: Boolean,
+    modifier: Modifier = Modifier,
+) {
+    // [START android_compose_pip_set_source_rect]
+    val context = LocalContext.current
+
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+        val pipModifier = modifier.onGloballyPositioned { layoutCoordinates ->
+            val builder = PictureInPictureParams.Builder()
+            if (shouldEnterPipMode) {
+                val sourceRect = layoutCoordinates.boundsInWindow().toAndroidRectF().toRect()
+                builder.setSourceRectHint(sourceRect)
+            }
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+                builder.setAutoEnterEnabled(shouldEnterPipMode)
+            }
+            context.findActivity().setPictureInPictureParams(builder.build())
+        }
+    }
+    // [END android_compose_pip_set_source_rect]
+}
+
+@Composable
+fun VideoPlayer2(
+    shouldEnterPipMode: Boolean,
+    modifier: Modifier = Modifier,
+) {
+    // [START android_compose_pip_set_aspect_ratio]
+    val context = LocalContext.current
+
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+        val pipModifier = modifier.onGloballyPositioned { layoutCoordinates ->
+            val builder = PictureInPictureParams.Builder()
+
+            if (shouldEnterPipMode) {
+                val sourceRect = layoutCoordinates.boundsInWindow().toAndroidRectF().toRect()
+                builder.setSourceRectHint(sourceRect)
+                builder.setAspectRatio(
+                    Rational(sourceRect.width(), sourceRect.height())
+                )
+            }
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+                builder.setAutoEnterEnabled(shouldEnterPipMode)
+            }
+            context.findActivity().setPictureInPictureParams(builder.build())
+        }
+    }
+    // [END android_compose_pip_set_aspect_ratio]
+}
+
+// [START android_compose_pip_build_remote_action]
+@RequiresApi(Build.VERSION_CODES.O)
+private fun buildRemoteAction(
+    @DrawableRes iconResId: Int,
+    @StringRes titleResId: Int,
+    requestCode: Int,
+    controlType: Int,
+    context: Context
+): RemoteAction {
+    return RemoteAction(
+        Icon.createWithResource(context, iconResId),
+        context.getString(titleResId),
+        context.getString(titleResId),
+        PendingIntent.getBroadcast(
+            context,
+            requestCode,
+            Intent(ACTION_BROADCAST_CONTROL)
+                .putExtra(EXTRA_CONTROL_TYPE, controlType),
+            PendingIntent.FLAG_IMMUTABLE
+        )
+    )
+}
+// [END android_compose_pip_build_remote_action]
+
+// [START android_compose_pip_broadcast_receiver]
+@RequiresApi(Build.VERSION_CODES.O)
+@Composable
+fun BroadcastReceiver(player: Player?) {
+    if (isInPipMode() && player != null) {
```

Comment:
Recommendation: This would be more readable if you do an early return. And perhaps log a warning statement:

```
    if (!isInPipMode() || player == null) {
         // Log statement
    }

    // ...rest of the code
```